Update SSMProblems.jl interface

THargreaves opened the pull request:
An update to the SSMProblems.jl interface (changes to documentation and examples to come in a separate PR).

The key changes are as follows:

- SSM dynamics are split into `LatentDynamics` and `ObservationProcess`
- A distribution can be defined which automatically implements both `transition` and `transition_logdensity` (likewise for initialisation and observations)
- A method for forward simulation of models is added
- States, time steps, and extras are passed in as keyword arguments which may be optional depending on the model definition
- **[Breaking]** The signature of the `transition_logdensity` method has been change from containing "previous"/"current" state to "current"/"next" state to make it consistent with `transition` (motivated by the off-by-one error in the PGAS implementation)

After discussing last week that we could pass `extras` in as a kwarg rather than as part of the conditional model, I realised that we could do the same with `observations`, making `ConditionedStateSpaceModel` redundant, hence it is not included in this PR.

A few features that I have not implemented in this PR yet in order to keep it simple:
- Optional RNGs
- Any dimension validation (e.g. making sure latent dynamics and observation process have matching dimensions)
- I have removed the particle container utility code to be replaced with a more general `AbstractParticleContainer`

A few things I wasn't sure about:
- Do we use US or UK spellings for method names? (initialise or initialize?)
- Is there a type stability issue with the definitions of methods for `StateSpaceModel`?
- Is using `hasmethod` an efficient way to check if a transition distribution is defined? Would a try-catch or cached value be better?

FredericWantiez commented on src/utils/particles.jl:
```diff
@@ -1,43 +0,0 @@
-"""
-module Utils
-
-abstract type Node{T} end
```
This should probably be in `AdvancedPS` but in the meantime, we need to move it to the `smc` example script that uses it I.

FredericWantiez commented on src/utils/distributions.jl:
```diff
@@ -0,0 +1,58 @@
+"""
+
+function check_initial_distribution(dynamics::LatentDynamics)
+    if !hasmethod(initialisation_distribution, Tuple{typeof(dynamics),Integer})
```
Why not defining a fallback implementation instead ? Or am I missing something ?
```julia
function initialisation_distribution(process::LatentDynamics, ::Integer)
    error("Method not defined for $(typeof(process))")
end
```

THargreaves replied:
My thought process behind this was trying to avoid a confusing error message.

Suppose the user implements `initial_logdensity` with the wrong signature. Then the particle filter will default to using `initial_logdensity` instead, trying to call the non-existant `initialisation_distribution`.

The error will make no mention of the fact that `initial_logdensity` is not defined, just that `initialisation_distribution` isn't.

This might be overkill though, and I'm not sure what the performance over head is.

FredericWantiez commented on examples/kalman-filter/script.jl:
```diff
@@ -1,43 +1,143 @@
+        x_pred, P_pred = if t == 1
+            z, P
+        else
+            Φ * x_filts[t - 1] + b, Φ * P_filts[t - 1] * Φ' + Q
```
We actually never use the SSMproblem API, maybe here we should be using `transition` ?

THargreaves replied:
Yes, I agree that this is a bit weird. 

We still do use `transition` if you take this model and run a particle filter on it (i.e. the `calc_` methods are used to define `transition`.

The difficulty of defining a `transition` method for the Kalman filtering case is that it breaks the separation of model definition and algorithm definition—the user has to start defining parts of the Kalman filtering algorithm rather than the raw input components (the 7 or so model matrices/vectors).

The `transition` function would have to take a `μ_filt` and `Σ_filt` and return updated `μ_pred` and `Σ_pred`. This isn't too bad, but it does require the user to write line 115 (`Φ * x_filts[t - 1] + b, Φ * P_filts[t - 1] * Φ' + Q`) by themselves.

It's then unclear what we would do for the filtering. This looks like this

```
K = P_pred * H' / (H * P_pred * H' + R)
x_filt = x_pred + K * (y - H * x_pred)
P_filt = P_pred - K * H * P_pred
```

which isn't in the form of a usual `observation_logdensity` call.

If we could figure out a way around the second issue though (perhaps an `update` method) that could look quite elegant.

We could then get around the model/algorithm separation issue by having an outer interface where the users define the `calc_` methods which then get used in a specific `transition(x::GaussianDistribution, ...` method defined in `AnalyticFilters.jl` (or whatever we end up calling it).

@charlesknipp, we talked about some similar ideas when we met (the idea that `transition` could pass around arbitrary representations of a filtering distribution—e.g. Gaussian, Dirichlet, collection of particles, collection of Gaussians). Any thoughts on this?

THargreaves replied:
Continuing this conversation in #39

THargreaves commented on examples/kalman-filter/script.jl:
```diff
@@ -1,43 +1,143 @@
+
+    for t in 1:T
+        # Extract parameters
+        z = calc_z(model.latent_dynamics, t, extras[t])
```
In this specific case, yes, since the model matrices/vectors are fixed over time. But in general, the A matrix might vary depending on the `extras` (e.g. when this is the inner model of a Rao-Blackwellised model). 

What we can do however is have a single function call `calc_params` which returns all of these matrices and vectors at once, which defaults to calling the individual `calc_XYZ` functions.

This is also important for efficient code, e.g. for the Lévy SSM, A and b can be computed together.

THargreaves commented on src/SSMProblems.jl:
```diff
@@ -1,40 +1,203 @@
+abstract type ObservationProcess end
+
+"""
+    observation(
```
This is a fair point. To be consistent with `transition_logdensity`/`transition`, my instinct would be to call these `observation_logdensity`/`observe`.

Subtle difference, but this now suggests that you are observing the process by generating an observation rather than doing something with an existing observation (which is what `update` would do).

If you still think that is still confusing though, `simulate_observation` works too.

THargreaves replied:
On reflection, for this simple example, there is probably no harm in unpacking, just to keep the code complexity low. I will make that change.

THargreaves replied:
I see no harm in `observation` being there. It is used for forward simulation of the model. 

I've updated the introductory tutorial to be defined in terms of `observation_distribution` so the user doesn't have to be aware of `observe` if they don't want to.

charlesknipp replied:
I see no problem with having `observe` as part of the interface. I personally use it in forecasting exercises, so it would be a nice feature to have. Although if we want to go that route, then it might be useful to have a `forward_simulation`-like method generate `x[t+1:t+k]` and `y[t+1:t+k]` for some `k` horizons ahead

THargreaves replied:
Yes, forward simulation was the main motivation for `observe`. There is currently a helper that simulates data for a fixed time `T` from the initialisation, but I could see something more general like you describe being very useful.

FredericWantiez replied:
I also think something like `simulate(obs::ObservationProcess, ...)` would be quite clear on the intent

THargreaves replied:
My only concern with this is that it introduces inconsistency into the API.

A SSM is governed by three distributions: initialisation, transition, and observation/emission.

Currently we define a "verb" form of each of these for sampling (e.g. initialise) and a log density function.

We also then have the distribution definition methods for convenience.

Keeping this breakdown for initialisation and transition but not observation feels a bit inconsistent.

The two consistent options are:

1. Keep things as they are, potentially replacing `observation` with another word. E.g. we could go back to `emission_distribution`, `emission_logdensity`, `emit`

2. We use `simulate` (or `rand`/`sample`) but we use it for all three distributions. That is, you have `simulate(dyn::LatentDynamics, step::Int, ...)` and `simulate(obs::ObservationProcess, step::Int, ...)` but this would mean we'd have to return to defining the initialisation distribution as `simulate(dyn::LatentDynamics, step::Int, ...)` (i.e. without `step`). I think this means we'd then need to have `distribution(dyn::LatentDynamics, step::Int, ...)` etc. for the two-in-one logdensity/simulation definition

THargreaves replied:
Okay, can do. Quite a big change so just checking if @FredericWantiez has any thoughts?

FredericWantiez replied:
Would that mean that we would only get one function ? `simulate`:

```julia
simulate(::LatentDynamics).# initialization
simulate(::LatentDynamics, step, ...) # transition
simulate(::ObservationDynamics, step, ...) # emission / observation
```

charlesknipp replied:
> ```julia
> simulate(::LatentDynamics).# initialization
> 
> simulate(::LatentDynamics, step, ...) # transition
> ```

I think this is a really smart design choice. As long as controls don't break this pattern.

charlesknipp commented on src/SSMProblems.jl:
```diff
@@ -1,40 +1,203 @@
 
 using AbstractMCMC: AbstractMCMC
+import Random: AbstractRNG
+
```
```suggestion
import Distributions: logpdf

```

otherwise `observation_logdensity()` will error

charlesknipp reviewed (commented):
It's pretty minor, but there is a type instability when using `PDMat{T}` for the linear dynamics/observation processes. Since the constructors require `Matrix`, you could just change the types to `PDMat{T, Matrix{T}}` without any issues

THargreaves commented:
Good point. I'll just remove the `PDMat` entirely. Seems overkill for what is meant to be an introductory example.

THargreaves commented:
I'm a bit at a loss as to why the Kalman filter example doc page keeps failing:

~~~
ERROR: LoadError: LoadError: ParseError:
# Error @ /Users/tim/Particle Filtering/SSMProblems.jl/docs/src/examples/kalman-filter.md:10:52
    x_filts = Vector{Vector{latent_type}}(undef, T)
    P_filts = Vector{Matrix{latent_type}}(undef, T)
#                                                  └ ── Expected `end`
in expression starting at /Users/tim/Particle Filtering/SSMProblems.jl/docs/src/examples/kalman-filter.md:10
when executing the following code block from inputfile `~/Particle Filtering/SSMProblems.jl/examples/kalman-filter/script.jl`

```julia
function AbstractMCMC.sample(
    model::LinearGaussianSSM,
    ::KalmanFilter,
    observations::AbstractVector,
    extras::AbstractVector,
)
    T = length(observations)
    latent_type = only(typeof(model.latent_dynamics).parameters)
    x_filts = Vector{Vector{latent_type}}(undef, T)
    P_filts = Vector{Matrix{latent_type}}(undef, T)
```
~~~

Does Literate.jl not allow blank lines in functions or am I missing something obvious?

FredericWantiez commented on examples/kalman-filter/script.jl:
```diff
@@ -1,44 +1,148 @@
-function transition!!(rng::AbstractRNG, model::LinearGaussianSSM, state::Gaussian)
-    let Φ = model.Φ, Q = model.Q, μ = state.μ, Σ = state.Σ
-        return Gaussian(Φ * μ, Φ * Σ * Φ' + Q)
+    @unpack z, P, Φ, b, Q = model.latent_dynamics # Extract parameters
```
Try ## for comments ?

FredericWantiez replied:
@yebai @THargreaves sorry for the spam, but the doc should build in [dbf3dbd](https://github.com/TuringLang/SSMProblems.jl/pull/38/commits/dbf3dbd0d5b36a4d395720f66557283b56ccb272)

THargreaves replied:
Brilliant. Thank you for taking a look.

THargreaves commented on examples/kalman-filter/script.jl:
```diff
@@ -1,44 +1,148 @@
+        x_pred, P_pred = if t == 1
+            z, P
+        else
+            Φ * x_filts[t - 1] + b, Φ * P_filts[t - 1] * Φ' + Q  ## Prediction step
```
That works. Though it is worth noting that the user will still have to define `calc_XYZ` for all matrices/vectors even if they define their `transition` function, if they want it to work with more general algorithms.

For example, the prediction/update step of square-root Kalman filter involves taking the QR decomposition of a block matrix made up of various A, Q, H, R, etc. which can't naturally be represented using the `transition` call.

For that reason, I suggest we define `transition` on the user's before for linear Gaussian models so that it just becomes an internal helper to keep the KalmanFilter code shorter rather than something the users try to implement themselves.

THargreaves replied:
I think the trouble with that approach is that the exact extras you need depend on the filtering algorithm you're using, so that the transition method stops being algorithm-agnostic.

I think for now, using `transition` where natural and not worrying about it otherwise is the least complicated approach. Not hard to add in an abstraction later if we realise there is a common thread between all the `update` methods.

THargreaves replied:
This change is probably outside of the scope of this PR since it relies on the more general transition functions (ones that can act on different types of states) discussed in #39. Will resolve this comment for now and can discuss there.

THargreaves commented:
Simplified the interface to only contain `simulate`, `logdensity`, `distribution`. Docs and example updated. Think we're ready to merge now.

In the process I removed the SSM level versions of the old functions, since I realised they were largely redundant, and instead it makes more sense to define just a forward simulation method for each SSM type, which I've now added.

One thing to note is that the VSCode Julia extension doesn't play well with the function overloading and displays numerous "Possible method call error."s. I think this is more of a bug with the parser/linter and also occurred with the old-old interface, before splitting `transition` into two functions.

FredericWantiez commented on src/SSMProblems.jl:
```diff
@@ -1,40 +1,185 @@
-    emission_logdensity(model, state, observation[, timestep, cache])
 
-Compute the log potential of the current particle. This effectively "reweight" each particle.
+function observation_logdensity end
```
Why do we still need this one ? Or is it just an alias to `logdensity(::ObservationProcess, ...)` ?

THargreaves replied:
Oh this is a silly mistake on my behalf. Have corrected now.

THargreaves commented on examples/kalman-filter/script.jl:
```diff
@@ -1,44 +1,146 @@
+
+const LinearGaussianSSM = StateSpaceModel{
+    <:LinearGaussianLatentDynamics,<:LinearGaussianObservationProcess
+}
```
Nice suggestion. Will update.
